### ISSUE TITLE
How to continue training from the last checkpoint

### ISSUE THREAD

**Issue by LinuxBeginner** [EdinburghNLP/nematus#116]
Hi, I am training the model in colab, so I will need to retrained the model. 
I used the following training script to train the model:

```
THEANO_FLAGS=mode=FAST_RUN,floatX=float32,device=$device,gpuarray.preallocate=0.8 python $nematus_home/nematus/train.py \
    --model $working_dir/model.npz \
    --datasets $data_dir/train.bpe.$src $data_dir/train.bpe.$trg \
    --valid_datasets $data_dir/dev.bpe.$src $data_dir/dev.bpe.$trg \
    --dictionaries $data_dir/train.bpe.$src.json $data_dir/train.bpe.$trg.json \
    --valid_script $script_dir/validate.sh \
    --dim_word 512 \
    --dim 1024 \
    --lrate 0.0001 \
    --optimizer adam \
    --maxlen 50 \
    --batch_size 80 \
    --valid_batch_size 40 \
    --validFreq 10000 \
    --dispFreq 1000 \
    --saveFreq 10000 \
    --sampleFreq 10000 \
    --tie_decoder_embeddings \
    --layer_normalisation \
    --dec_base_recurrence_transition_depth 8 \
    --enc_recurrence_transition_depth 4
```

After a while, the training stopped because I used my daily 12 hours limit in colab. I was able to save only one model (model.npz.data-00000-of-00001): 

```
enmn.bpe		       model.npz.json		truecase-model.en
model.npz.data-00000-of-00001  model.npz.meta		truecase-model.mn
model.npz.index		       model.npz.progress.json
```

As per the comments [here](https://github.com/marian-nmt/marian/issues/224#issuecomment-443348086) the parameter `--model $working_dir/model.npz` should do the work.

So, without any changes, when I trained the above script, the training starts again from epoch 0.

I also tried by adding  
`--reload $working_dir/model.npz.data-00000-of-00001` in the script, but it gave me an error.

which parameter should I use to continue training from the last checkpoint?

**Comment by rsennrich** [comment]
using "--reload $working_dir/model.npz" should do the trick.

On a side note, make sure to save checkpoints frequently enough ("--save_freq" and maybe "--valid_freq as well") that you're not wasting a lot of compute.

**Comment by rsennrich** [comment]
yes, you can do domain adaptation with the same command, simply switching out the ``--datasets`` (and ``--valid_datasets``) parameters.

no, there is no need to add other parameters when switching out the corpus.

**Comment by LinuxBeginner** [comment]
Hi, @rsennrich thank you for the quick reply.

About my earlier query to continue training from the last checkpoint. 

When I trian for the 1st time, it went up to 10 epoch, when I retrain it again, it started from 6th epoch.
1st train==>  1-10 epoch

I think this might be because of the value for "--save_freq" was large.
After the second run went for 20 epoch and the system stopped. 

2nd retrain==>  6-20 epoch
On retraining again it started from 14th epoch. Fine.

3rd retrain==> 14-21 epoch 

But, on the fourth retrain, it again started from 14th epoch, what could be the reason?
4th retrain==> 14 ....

The `checkpoint` file seem to record the latest model saved in it 

```
model_checkpoint_path: "model.npz-15000"
all_model_checkpoint_paths: "model.npz-15000"
```

Shouldn't the system pick up the last epoch on which the model was saved, let's say the training went up to 30th epoch, the last saved model was in 25th epoch, the system should start from the 25th one on retrain right?

**Comment by LinuxBeginner** [comment]
Hi @rsennrich ,

So, I train the system from the start, with the following script

```
THEANO_FLAGS=mode=FAST_RUN,floatX=float32,device=$device,gpuarray.preallocate=0.8 python $nematus_home/nematus/train.py \
    --model $working_dir/model.npz \
    --datasets $data_dir/train.bpe.$src $data_dir/train.bpe.$trg \
    --valid_datasets $data_dir/dev.bpe.$src $data_dir/dev.bpe.$trg \
    --dictionaries $data_dir/train.bpe.$src.json $data_dir/train.bpe.$trg.json \
    --dim_word 512 \
    --dim 1024 \
    --lrate 0.0001 \
    --optimizer adam \
    --maxlen 50 \
    --batch_size 80 \
    --valid_batch_size 40 \
    --validFreq 2000 \
    --dispFreq 1000 \
    --saveFreq 4000 \
    --sampleFreq 5000 \
    --tie_decoder_embeddings \
    --layer_normalisation \
    --dec_base_recurrence_transition_depth 8 \
    --enc_recurrence_transition_depth 4

```

it ran up to 65 epoch. ( `Validation cross entropy (AVG/SUM/N_SENTS/N_TOKENS): 104.05570943714682 45680.45644290745 439 12761` )
On retrainig it, the script seems to pick up the epoch where the average value **AVG** of
`Validation cross entropy (AVG/SUM/N_SENTS/N_TOKENS):`
was the lowest, which in my case was at epoch 21 
`Validation cross entropy (AVG/SUM/N_SENTS/N_TOKENS): 93.52767660609682 41058.650030076504 439 12761`

So, on retrain, the system starts to train from epoch 21. Is there any way where I could retrain from the latest save checkpoint instead of lowest  **AVG** of Validation cross entropy?

**Note**: I am training the system for a low resource language with less than 30k corpus.

**Comment by rsennrich** [comment]
model.npz is generally the checkpoint with the lowest validation perplexity. If you want to continue training from the latest checkpoint, you can use ``--reload model.npz-XXXX`` (where XXXX is the number of the latest checkpoint).

if you train a base model and then continue fine-tuning it on some other data, I also find it a good practice to change the working directory (and possibly the model name) so I don't unintentionally overwrite the base model.

**Comment by LinuxBeginner** [comment]
Thank you @rsennrich  for the answer.

The lower the value of **Validation cross entropy AVG** the better is the system.

The one where I train the system from scratch again, AVG value kept on increasing. What does the **Loss/word** represent?

If the increase of **Validation cross entropy AVG** meant, my system is already converge, then why does the value of **Loss/word** kept on decreasing?

I also check the output of translate.py with model.npz-4000 and model.npz-12000. Both are same.
